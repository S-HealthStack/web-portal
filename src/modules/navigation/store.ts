import { createSelector } from '@reduxjs/toolkit';
import { matchPath } from 'react-router-dom';
import { createBrowserHistory, createMemoryHistory, Location } from 'history';

import { RootState } from 'src/modules/store/store';

export enum Path {
  Root = '/',

  SignIn = '/signin',
  AccountActivation = '/account/activation',
  AccountCreate = '/account/create',
  AccountConfirm = '/account/confirm',
  AccountVerification = '/email-verification',
  ForgotPassword = '/account/forgot',
  ForgotPasswordConfirm = '/account/forgot/confirm',
  ResetPassword = '/account/forgot/reset',
  ResetPasswordComplete = '/account/forgot/complete',

  CreateStudy = '/create-study',

  Overview = '/overview',
  OverviewSubject = '/overview/participants/:subjectId',
  TrialManagement = '/trial-management',
  TrialManagementSubject = '/trial-management/participants/:subjectId',
  TrialManagementEditSurvey = '/trial-management/surveys/:surveyId/edit',
  TrialManagementSurveyResults = '/trial-management/surveys/:surveyId/results',
  UserAnalytics = '/user-analytics',
  DataCollection = '/data-collection',
  DataCollectionSubject = '/data-collection/participants/:subjectId',
  StudySettings = '/study-settings',
}

export const makeHistory = () =>
  process.env.NODE_ENV !== 'test'
    ? createBrowserHistory({ basename: process.env.PUBLIC_PATH })
    : createMemoryHistory();

export const history = makeHistory();

const sectionPaths = [
  Path.Overview,
  Path.OverviewSubject,
  Path.TrialManagement,
  Path.UserAnalytics,
  Path.DataCollection,
  Path.StudySettings,
] as const;

export const sectionPathSelector = createSelector(
  (state: RootState) => state.router,
  (r) => sectionPaths.find((s) => matchPath((r.location as unknown as Location).pathname, s))
);
