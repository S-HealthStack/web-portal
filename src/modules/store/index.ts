export type { RootState, AppDispatch, AppThunk } from './store';
export { useAppDispatch, useAppSelector } from './hooks';
export * from './tools';
