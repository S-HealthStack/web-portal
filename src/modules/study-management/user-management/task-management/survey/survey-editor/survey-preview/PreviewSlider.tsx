import React, { FC, useMemo } from 'react';

import _range from 'lodash/range';
import styled from 'styled-components';

import { colors, px, typography } from 'src/styles';

const POINTER_SIZE = 12;

const SliderContainer = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
  position: relative;
  height: ${px(70)};
`;

const SliderBar = styled.div`
  width: 100%;
  background-color: ${colors.primaryLight};
  height: ${px(2)};
`;

const ScaleWrapper = styled.div`
  width: 100%;
  height: fit-content;
  position: absolute;
  display: flex;
  align-items: center;
  justify-content: flex-start;
  padding: 0 ${px(POINTER_SIZE / 2)};
`;

const SliderScalePointWrapper = styled.div<{ index: number; areasNumber: number }>`
  cursor: pointer;
  height: ${px(36)};
  position: relative;
  top: ${px(-5)};
  display: flex;
  justify-content: ${({ index, areasNumber }) => {
    switch (index) {
      case 0:
        return 'flex-start';
      case areasNumber - 1:
        return 'flex-end';
      default:
        return 'center';
    }
  }};
  width: ${({ areasNumber, index }) => {
    if (index === 0 || index === areasNumber - 1) {
      return `calc((100% / ${areasNumber - 1}) / 2)`;
    }
    return `calc(100% / ${areasNumber - 1})`;
  }};
`;

const SliderScalePoint = styled.div`
  cursor: pointer;
  width: ${px(2)};
  position: relative;
  top: ${px(20)};
  height: ${px(6)};
  background-color: ${colors.primary};
`;

const ValuesContainer = styled.div`
  width: 100%;
  display: flex;
  justify-content: space-between;
  ${typography.labelRegular};
  line-height: ${px(13)};
  color: ${colors.textPrimary};
  position: absolute;
  top: ${px(28)};
  left: 0;
`;

const LabelsContainer = styled(ValuesContainer)`
  color: ${colors.textSecondaryGray};
  text-transform: uppercase;
  letter-spacing: ${px(1)};
  top: ${px(45)};
`;

const Pointer = styled.div<{ index?: number; areasNumber: number }>`
  position: absolute;
  width: ${px(16)};
  height: ${px(16)};
  background-color: ${colors.primary};
  border-radius: 50%;
  box-shadow: 0 0 ${px(8)} rgba(32, 87, 213, 0.25); // TODO unknown color
  top: ${px(-8)};
  left: ${({ index, areasNumber }) => {
    if (index === 0) {
      return '0';
    }
    if (index === areasNumber - 1) {
      return `calc(100% - ${px(POINTER_SIZE)} - ${px(4)})`;
    }
    return `calc(((100% - ${px(POINTER_SIZE)} - ${px(4)}) / (${areasNumber} - 1) * ${index || 0}))`;
  }};
  opacity: ${({ index }) => (index !== undefined ? 1 : 0)};
`;

interface PreviewSliderProps {
  maxIndex: number;
  minIndex: number;
  maxLabel: string;
  minLabel: string;
  activeIndex?: number;
  onChange: (index: number) => void;
}

const PreviewSlider: FC<PreviewSliderProps> = ({
  maxIndex,
  minIndex,
  maxLabel,
  minLabel,
  activeIndex,
  onChange,
}: PreviewSliderProps) => {
  const areasNumber = maxIndex - minIndex + 1;

  const selectedPoint = useMemo(() => {
    if (minIndex === 0) {
      return activeIndex;
    }

    if (activeIndex) {
      return activeIndex - 1;
    }

    return undefined;
  }, [activeIndex, minIndex]);

  return (
    <SliderContainer data-testid="survey-preview-slider">
      <ScaleWrapper>
        {_range(0, areasNumber).map((i) => (
          <SliderScalePointWrapper
            data-testid="survey-preview-slider-point"
            index={i}
            areasNumber={areasNumber}
            onClick={() => onChange(i + minIndex)}
            key={i}
          >
            <SliderScalePoint />
          </SliderScalePointWrapper>
        ))}
      </ScaleWrapper>
      <SliderBar />
      <Pointer index={selectedPoint} areasNumber={areasNumber} />
      <ValuesContainer>
        <div data-testid="survey-preview-slider-min-value">{minIndex}</div>
        <div data-testid="survey-preview-slider-max-value">{maxIndex}</div>
      </ValuesContainer>
      <LabelsContainer>
        <div data-testid="survey-preview-slider-min-label">{minLabel}</div>
        <div data-testid="survey-preview-slider-max-label">{maxLabel}</div>
      </LabelsContainer>
    </SliderContainer>
  );
};

export default PreviewSlider;
