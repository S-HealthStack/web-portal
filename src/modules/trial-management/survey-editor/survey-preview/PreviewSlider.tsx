import styled from 'styled-components';
import React, { FC, useMemo } from 'react';
import _range from 'lodash/range';

import { colors, px, typography } from 'src/styles';

const SliderContainer = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
  position: relative;
`;

const SliderBar = styled.div`
  width: 100%;
  background-color: ${colors.updPrimaryLight};
  height: ${px(2)};
`;

const SliderScalePointWrapper = styled.div<{ index: number; areasNumber: number }>`
  cursor: pointer;
  width: ${({ areasNumber }) => `calc(100% / ${areasNumber})`};
  height: ${px(6)};
  position: absolute;
  left: ${({ index, areasNumber }) => `calc(100% / ${areasNumber} * ${index})`};
  top: ${px(14)};
  display: flex;
  justify-content: center;
`;

const SliderScalePoint = styled.div`
  cursor: pointer;
  width: ${px(2)};
  height: ${px(6)};
  background-color: ${colors.updPrimary};
`;

const ValuesContainer = styled.div`
  width: 100%;
  display: flex;
  justify-content: space-between;
  ${typography.labelRegular};
  line-height: ${px(13)};
  color: ${colors.updTextPrimary};
  position: absolute;
  top: ${px(28)};
  left: 0;
`;

const LabelsContainer = styled(ValuesContainer)`
  color: ${colors.updTextSecondaryGray};
  text-transform: uppercase;
  letter-spacing: ${px(1)};
  top: ${px(45)};
`;

const SelectedPoint = styled.div<{ index?: number; areasNumber: number }>`
  position: absolute;
  width: ${px(16)};
  height: ${px(16)};
  background-color: ${colors.updPrimary};
  border-radius: ${px(25)};
  box-shadow: 0 0 ${px(8)} rgba(32, 87, 213, 0.25);
  top: ${px(-8)};
  left: ${({ index, areasNumber }) =>
    `calc((100% / ${areasNumber} * ${index || 0}) + (100% / ${areasNumber} / 2) - ${px(8)})`};
  opacity: ${({ index }) => (index !== undefined ? 1 : 0)};
`;

interface PreviewSliderProps {
  maxIndex: number;
  minIndex: number;
  maxLabel: string;
  minLabel: string;
  activeIndex?: number;
  onChange: (index: number) => void;
}

const PreviewSlider: FC<PreviewSliderProps> = ({
  maxIndex,
  minIndex,
  maxLabel,
  minLabel,
  activeIndex,
  onChange,
}: PreviewSliderProps) => {
  const areasNumber = maxIndex - minIndex + 1;

  const selectedPoint = useMemo(() => {
    if (minIndex === 0) {
      return activeIndex;
    }

    if (activeIndex) {
      return activeIndex - 1;
    }

    return undefined;
  }, [activeIndex, minIndex]);

  return (
    <SliderContainer>
      <SliderBar />
      {_range(0, areasNumber).map((i) => (
        <SliderScalePointWrapper
          index={i}
          areasNumber={areasNumber}
          onClick={() => onChange(i + minIndex)}
        >
          <SliderScalePoint />
        </SliderScalePointWrapper>
      ))}
      <ValuesContainer>
        <div>{minIndex}</div>
        <div>{maxIndex}</div>
      </ValuesContainer>
      <LabelsContainer>
        <div>{minLabel}</div>
        <div>{maxLabel}</div>
      </LabelsContainer>
      <SelectedPoint index={selectedPoint} areasNumber={areasNumber} />
    </SliderContainer>
  );
};

export default PreviewSlider;
