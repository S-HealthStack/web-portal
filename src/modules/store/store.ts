/* eslint-disable import/no-import-module-exports */
import {
  Action,
  ThunkAction,
  configureStore,
  getDefaultMiddleware,
  ThunkDispatch,
} from '@reduxjs/toolkit';
import { routerMiddleware } from 'connected-react-router';

import { history } from 'src/modules/navigation/store';
import rootReducer from 'src/modules/store/reducers';

export { rootReducer };

export const store = configureStore({
  reducer: rootReducer,
  middleware: [...getDefaultMiddleware(), routerMiddleware(history)],
  devTools: process.env.NODE_ENV !== 'production',
});

if (module.hot) {
  // Enable Webpack hot module replacement for reducers
  module.hot.accept('./reducers', async () => {
    // eslint-disable-next-line import/no-useless-path-segments
    const nextRootReducer = (await import('../store/reducers')).default;
    store.replaceReducer(nextRootReducer);
  });
}

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = ThunkDispatch<RootState, unknown, Action>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
