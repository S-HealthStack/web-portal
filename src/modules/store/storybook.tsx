import React from 'react';
import { configureStore, ConfigureStoreOptions } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { DecoratorFunction } from '@storybook/csf';
import { ReactFramework } from '@storybook/react';
import { MemoryRouter } from 'react-router-dom';
import { rootReducer, RootState } from './store';

export const getStoreDecorator = (state: ConfigureStoreOptions<RootState>['preloadedState']) => {
  const store = configureStore({
    reducer: rootReducer,
    preloadedState: state,
  });

  return ((story) => (
    <Provider store={store}>
      <MemoryRouter>{story()}</MemoryRouter>
    </Provider>
  )) as DecoratorFunction<ReactFramework>;
};
