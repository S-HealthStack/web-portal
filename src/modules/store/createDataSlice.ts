import { useCallback, useEffect } from 'react';
import _isEqual from 'lodash/isEqual';
import useMount from 'react-use/lib/useMount';
import {
  ActionCreatorWithoutPayload,
  ActionCreatorWithPayload,
  createSlice,
  PayloadAction,
} from '@reduxjs/toolkit';
import type { Reducer } from 'redux';

import applyDefaultApiErrorHandlers from 'src/modules/api/applyDefaultApiErrorHandlers';
import { SnackbarInstance } from 'src/modules/snackbar/snackbar.slice';
import useShowSnackbar from 'src/modules/snackbar/useShowSnackbar';
import { AppThunk, ErrorType, RootState, WithLoadableData } from 'src/modules/store';

import { useAppDispatch, useAppSelector } from './hooks';

type PublicDataSliceState<D> = WithLoadableData<D>;

type EmptyFetchArgs = null;

export type DataSliceState<D, A> = PublicDataSliceState<D> & {
  fetchArgs: A | EmptyFetchArgs;
  prevFetchArgs: A | EmptyFetchArgs;
  isLoadingSilently?: boolean;
};

type DataSliceOptions<D, N extends string, A> = {
  name: N;
  fetchData: (args: A, dispatch: unknown) => Promise<D>;
};

type FetchActionOptions = {
  force?: boolean;
  silent?: boolean;
  onError?: (e?: unknown) => void;
};

type HookParams<A, D> = {
  initialData?: D;
  fetchArgs: A | boolean;
  refetchSilentlyOnMount?: boolean;
};

export type ShowOnErrorSnackbarParams = Omit<SnackbarInstance, 'id' | 'hideTimeoutId'> & {
  onAction?: () => void;
};

type DataSliceReturnType<D, N, A> = {
  name: N;
  reducer: Reducer<DataSliceState<D, A>>;
  stateSelector: (s: RootState) => DataSliceState<D, A>;
  hook: (
    params?: HookParams<A, D>,
    snackBarParams?: ShowOnErrorSnackbarParams
  ) => DataSliceState<D, A> & {
    fetch: (args: A, opts?: FetchActionOptions) => Promise<void>;
    refetch: () => Promise<void>;
    reset: () => void;
  };
  actions: {
    reset: ActionCreatorWithoutPayload;
    setData: ActionCreatorWithPayload<{ data: D }>;
    loadingStarted: ActionCreatorWithPayload<{ fetchArgs: A }>;
    loadingSuccess: ActionCreatorWithPayload<{ data: D }>;
    loadingFailed: ActionCreatorWithPayload<ErrorType>;
    fetch: (args: A, opts?: FetchActionOptions) => AppThunk<Promise<void>>;
    refetch: () => AppThunk<Promise<void>>;
  };
};

const isFetchArgsEmpty = <D, A>(
  fetchArgs: DataSliceState<D, A>['fetchArgs']
): fetchArgs is EmptyFetchArgs => fetchArgs === null;

function createDataSlice<D extends object, N extends string, A = void>(
  options: DataSliceOptions<D, N, A>
): DataSliceReturnType<D, N, A> {
  const initialState: DataSliceState<D, A> = {
    fetchArgs: null,
    prevFetchArgs: null,
  };

  const slice = createSlice({
    name: options.name,
    initialState,
    reducers: {
      setData(state, action: PayloadAction<{ data: D }>) {
        return Object.assign(state, {
          data: action.payload.data,
        });
      },
      loadingStarted(state, action: PayloadAction<{ fetchArgs: A; silent?: boolean }>) {
        return Object.assign(state, {
          isLoading: !action.payload.silent,
          isLoadingSilently: !!action.payload.silent,
          fetchArgs: action.payload.fetchArgs,
          prevFetchArgs: { ...state.fetchArgs },
          error: undefined,
        });
      },
      loadingSuccess(state, action: PayloadAction<{ data: D }>) {
        return Object.assign(state, {
          isLoading: false,
          isLoadingSilently: false,
          error: undefined,
          data: action.payload.data,
          prevFetchArgs: { ...state.fetchArgs },
        });
      },
      loadingFailed(state, action: PayloadAction<ErrorType>) {
        return Object.assign(state, {
          isLoading: false,
          isLoadingSilently: false,
          error: String(action.payload),
          data: undefined,
          prevFetchArgs: { ...state.fetchArgs },
        });
      },
      reset(state) {
        state.isLoading = undefined;
        state.isLoadingSilently = undefined;
        state.data = undefined;
        state.error = undefined;

        return Object.assign(state, initialState);
      },
    },
  });

  const sliceStateSelector = (state: RootState) =>
    state[options.name as keyof RootState] as DataSliceState<D, A>;

  const fetchThunk =
    (args: A, opts?: FetchActionOptions): AppThunk<Promise<void>> =>
    async (dispatch, getState) => {
      if (!opts?.force) {
        const { fetchArgs: currentArgs } = sliceStateSelector(getState());

        // fetch only if this is a first fetch or args have been changed
        if (!isFetchArgsEmpty(currentArgs) && _isEqual(currentArgs, args)) {
          return;
        }
      }
      dispatch(slice.actions.loadingStarted({ fetchArgs: args, silent: opts?.silent }));
      try {
        const data = await options.fetchData(args, dispatch);

        const { fetchArgs: currentArgs } = sliceStateSelector(getState());
        // discard received data
        // if another fetch with different args performed while fetchData was in progress
        if (_isEqual(args, currentArgs)) {
          dispatch(slice.actions.loadingSuccess({ data }));
        }
      } catch (err) {
        const { fetchArgs: currentArgs } = sliceStateSelector(getState());
        if (_isEqual(args, currentArgs)) {
          console.error(err);
          dispatch(
            slice.actions.loadingFailed(
              (err instanceof Error ? err.message : undefined) ||
                `Failed to fetch data ${options.name}`
            )
          );

          if (!applyDefaultApiErrorHandlers(err, dispatch, false)) {
            opts?.onError?.(err);
          }
        }
      }
    };

  const refetchThunk = (): AppThunk<Promise<void>> => async (dispatch, getState) => {
    const { fetchArgs } = sliceStateSelector(getState());
    if (!isFetchArgsEmpty(fetchArgs)) {
      await dispatch(fetchThunk(fetchArgs, { force: true }));
    }
  };

  const refetchSilentlyIfRequiredThunk =
    (): AppThunk<Promise<void>> => async (dispatch, getState) => {
      const state = sliceStateSelector(getState());
      if (!state.isLoading && !state.isLoadingSilently && !isFetchArgsEmpty(state.fetchArgs)) {
        await dispatch(fetchThunk(state.fetchArgs, { force: true, silent: !state.error }));
      }
    };

  const useSliceHook = (params?: HookParams<A, D>, snackBarParams?: ShowOnErrorSnackbarParams) => {
    const dispatch = useAppDispatch();
    const sliceState = useAppSelector((state) => sliceStateSelector(state));
    const isStudyListLoading = useAppSelector((state) => state.studies.isLoading);
    const showSnackBar = useShowSnackbar();

    const onError = () => {
      if (snackBarParams) {
        showSnackBar(snackBarParams);
      }
    };

    const hookResult = {
      fetch: useCallback(
        (args: A, opts?: FetchActionOptions) => dispatch(fetchThunk(args, { ...opts, onError })),
        // eslint-disable-next-line react-hooks/exhaustive-deps
        [dispatch]
      ),
      refetch: useCallback(() => dispatch(refetchThunk()), [dispatch]),
      reset: useCallback(() => dispatch(slice.actions.reset()), [dispatch]),
      ...sliceState,
      isLoading: sliceState.isLoading || isStudyListLoading,
    };

    const hasParams = !!params;
    useEffect(() => {
      if (hasParams && typeof params.fetchArgs !== 'boolean') {
        dispatch(fetchThunk(params.fetchArgs, { onError }));
      }
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [dispatch, hasParams, params?.fetchArgs]);

    useEffect(() => {
      if (
        !params?.initialData ||
        (!isFetchArgsEmpty(sliceState.fetchArgs) &&
          _isEqual(sliceState.fetchArgs, params?.fetchArgs))
      ) {
        return;
      }

      dispatch(slice.actions.setData({ data: params.initialData }));
    }, [dispatch, params?.fetchArgs, params?.initialData, sliceState.data, sliceState.fetchArgs]);

    useMount(() => {
      if (params?.refetchSilentlyOnMount) {
        dispatch(refetchSilentlyIfRequiredThunk());
      }
    });

    return hookResult;
  };

  return {
    name: options.name,
    reducer: slice.reducer,
    stateSelector: sliceStateSelector,
    hook: useSliceHook,
    actions: {
      ...slice.actions,
      fetch: fetchThunk,
      refetch: refetchThunk,
    },
  };
}

export default createDataSlice;
