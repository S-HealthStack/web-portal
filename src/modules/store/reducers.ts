import { combineReducers } from '@reduxjs/toolkit';
import { connectRouter } from 'connected-react-router';
import type { Reducer } from 'redux';

import authReducer from 'src/modules/auth/auth.slice';
import studySettingsReducer from 'src/modules/study-settings/studySettings.slice';
import snackbarReducer from 'src/modules/snackbar/snackbar.slice';
import overviewReducers from 'src/modules/overview/overview.slice';
import studiesReducer from 'src/modules/studies/studies.slice';
import studyManagementReducer from 'src/modules/study-management/studyManagement.slice';
import dataCollectionReducer from 'src/modules/data-collection/dataCollection.slice';
import sidebarReducer from 'src/modules/main-layout/sidebar/sidebar.slice';
import fileUploadReducer from 'src/modules/file-upload/fileUpload.slice';
import { history } from 'src/modules/navigation/store';
import { DataSliceState } from 'src/modules/store/createDataSlice';
import forgotPasswordReducers from 'src/modules/auth/forgot-password/forgotPassword.slice';
import resetPasswordReducers from 'src/modules/auth/forgot-password/resetPassword.slice';
import objectStorageReducers from 'src/modules/object-storage/objectStorage.slice';

const createReducer = <T, V>(asyncReducers?: { [key: string]: Reducer<DataSliceState<T, V>> }) =>
  combineReducers({
    ...authReducer,
    router: connectRouter(history),
    snackbar: snackbarReducer,
    dataCollection: dataCollectionReducer,
    sidebar: sidebarReducer,
    ...studySettingsReducer,
    ...overviewReducers,
    ...studyManagementReducer,
    ...forgotPasswordReducers,
    ...resetPasswordReducers,
    ...fileUploadReducer,
    ...asyncReducers,
    ...objectStorageReducers,
    ...studiesReducer,
  });

export default createReducer;
