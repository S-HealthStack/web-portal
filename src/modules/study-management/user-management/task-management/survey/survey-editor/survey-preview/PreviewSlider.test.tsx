import React from 'react';
import { act, render, screen } from '@testing-library/react';
import 'jest-styled-components';
import '@testing-library/jest-dom/extend-expect';
import { ThemeProvider } from 'styled-components';
import theme from 'src/styles/theme';
import { Provider } from 'react-redux';
import { ConnectedRouter } from 'connected-react-router';
import userEvent from '@testing-library/user-event';
import { makeStore } from 'src/modules/store/store';
import { makeHistory } from 'src/modules/navigation/store';
import PreviewSlider from './PreviewSlider';

describe('PreviewSlider', () => {
  let store: ReturnType<typeof makeStore>;
  let history: ReturnType<typeof makeHistory>;

  beforeEach(() => {
    history = makeHistory();
    store = makeStore(history);
  });

  it('should render', async () => {
    const minIndex = 1;
    const maxIndex = 10;
    const maxLabel = 'max';
    const minLabel = 'min';
    const activeIndex = 5;
    const onChange = jest.fn();

    await act(() => {
      render(
        <ThemeProvider theme={theme}>
          <Provider store={store}>
            <ConnectedRouter history={history}>
              <PreviewSlider
                minIndex={minIndex}
                maxIndex={maxIndex}
                maxLabel={maxLabel}
                minLabel={minLabel}
                activeIndex={activeIndex}
                onChange={onChange}
              />
            </ConnectedRouter>
          </Provider>
        </ThemeProvider>
      );
    });

    expect(await screen.findByTestId('survey-preview-slider')).toMatchSnapshot();

    const maxLabelElement = await screen.findByTestId('survey-preview-slider-max-label');
    const minLabelElement = await screen.findByTestId('survey-preview-slider-min-label');
    const maxValueElement = await screen.findByTestId('survey-preview-slider-max-value');
    const minValueElement = await screen.findByTestId('survey-preview-slider-min-value');
    const sliderPoints = await screen.findAllByTestId('survey-preview-slider-point');

    expect(maxLabelElement).toHaveTextContent(maxLabel);
    expect(minLabelElement).toHaveTextContent(minLabel);
    expect(maxValueElement).toHaveTextContent(String(maxIndex));
    expect(minValueElement).toHaveTextContent(String(minIndex));
    expect(sliderPoints).toHaveLength(maxIndex - minIndex + 1);

    await userEvent.click(sliderPoints[1]);
    expect(onChange).toHaveBeenCalled();
  });
});
