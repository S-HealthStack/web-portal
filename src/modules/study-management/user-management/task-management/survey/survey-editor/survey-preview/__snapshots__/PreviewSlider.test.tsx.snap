// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewSlider should render 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  position: relative;
  height: 4.375rem;
}

.c6 {
  width: 100%;
  background-color: #ECF1FC;
  height: 0.125rem;
}

.c1 {
  width: 100%;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  padding: 0 0.375rem;
}

.c2 {
  cursor: pointer;
  height: 2.25rem;
  position: relative;
  top: -0.3125rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  width: calc((100% / 9) / 2);
}

.c4 {
  cursor: pointer;
  height: 2.25rem;
  position: relative;
  top: -0.3125rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: calc(100% / 9);
}

.c5 {
  cursor: pointer;
  height: 2.25rem;
  position: relative;
  top: -0.3125rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  width: calc((100% / 9) / 2);
}

.c3 {
  cursor: pointer;
  width: 0.125rem;
  position: relative;
  top: 1.25rem;
  height: 0.375rem;
  background-color: #4475E3;
}

.c8 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  font-family: Inter,sans-serif;
  font-size: 0.625rem;
  line-height: 0.9375rem;
  font-weight: 400;
  font-style: normal;
  line-height: 0.8125rem;
  color: #474747;
  position: absolute;
  top: 1.75rem;
  left: 0;
}

.c9 {
  color: #808080;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.0625rem;
  -moz-letter-spacing: 0.0625rem;
  -ms-letter-spacing: 0.0625rem;
  letter-spacing: 0.0625rem;
  top: 2.8125rem;
}

.c7 {
  position: absolute;
  width: 1rem;
  height: 1rem;
  background-color: #4475E3;
  border-radius: 50%;
  box-shadow: 0 0 0.5rem rgba(32,87,213,0.25);
  top: -0.5rem;
  left: calc(((100% - 0.75rem - 0.25rem) / (10 - 1) * 4));
  opacity: 1;
}

<div
  class="c0"
  data-testid="survey-preview-slider"
>
  <div
    class="c1"
  >
    <div
      class="c2"
      data-testid="survey-preview-slider-point"
    >
      <div
        class="c3"
      />
    </div>
    <div
      class="c4"
      data-testid="survey-preview-slider-point"
    >
      <div
        class="c3"
      />
    </div>
    <div
      class="c4"
      data-testid="survey-preview-slider-point"
    >
      <div
        class="c3"
      />
    </div>
    <div
      class="c4"
      data-testid="survey-preview-slider-point"
    >
      <div
        class="c3"
      />
    </div>
    <div
      class="c4"
      data-testid="survey-preview-slider-point"
    >
      <div
        class="c3"
      />
    </div>
    <div
      class="c4"
      data-testid="survey-preview-slider-point"
    >
      <div
        class="c3"
      />
    </div>
    <div
      class="c4"
      data-testid="survey-preview-slider-point"
    >
      <div
        class="c3"
      />
    </div>
    <div
      class="c4"
      data-testid="survey-preview-slider-point"
    >
      <div
        class="c3"
      />
    </div>
    <div
      class="c4"
      data-testid="survey-preview-slider-point"
    >
      <div
        class="c3"
      />
    </div>
    <div
      class="c5"
      data-testid="survey-preview-slider-point"
    >
      <div
        class="c3"
      />
    </div>
  </div>
  <div
    class="c6"
  />
  <div
    class="c7"
  />
  <div
    class="c8"
  >
    <div
      data-testid="survey-preview-slider-min-value"
    >
      1
    </div>
    <div
      data-testid="survey-preview-slider-max-value"
    >
      10
    </div>
  </div>
  <div
    class="c8 c9"
  >
    <div
      data-testid="survey-preview-slider-min-label"
    >
      min
    </div>
    <div
      data-testid="survey-preview-slider-max-label"
    >
      max
    </div>
  </div>
</div>
`;
